man/man1/fz.1
.Dd 2024-10-26
.Dt fz 1
.Os
.Sh NAME
.Nm fz
.Nd A highly customisable monitoring daemon
.Sh SYNOPSIS
.Nm
.Op Fl n
.Op Fl c Ar file
.Op Fl l Ar error/warn/info/debug
.Sh DESCRIPTION
The
.Nm
daemon is a specialised task scheduler for monitoring. It raises alerts under specific conditions. The scripts themselves are typically shell scripts or small programs. They must adhere to a specific interface, but what they do specifically is up to you.
.Pp
Although some scripts are provided, you are expected and encouraged to write your own making
.Nm
infinitely flexible, though it does require a small amount of programming experience.
.Bl -tag -width Ds
.It Fl c, Li --config
The configuration file. See
.Xr flamingzombies.toml 5
for details on how it can be configured.
.It Fl C, Li --directory
Change to this directory before executing any script.
.It Fl h, Li --help
Basic usage information.
.It Fl l, Li --loglevel
The log level to set when running the daemon. Valid options are error, warn, info or debug.
.It Fl n, Li --configtest
Test the validity of the supplied configuration.
.Xr flamingzombies.toml 5
for details on how it can be configured.
.It Fl p, Li --pidfile
Path to write the daemons pidfile.
.It Fl V, Li --version
Print the version information.
.El
.Sh ENVIRONMENT
Most settings can also be supplied with environment variables, but if a command arguments is provided it will take precedence. The supported environment variables are listed below and closely match the arguments documented above.
.Bl -tag -width Ds
.It Ic FZ_CONFIG_FILE
.It Ic FZ_DIRECTORY
.It Ic FZ_LISTEN
.It Ic FZ_LOG_LEVEL
.El
.Sh SEE ALSO
.Xr fzctl 1 ,
.Xr flamingzombies.toml 5 ,
.Xr fz-gates 7 ,
.Xr fz-notifiers 7 ,
.Xr fz-tasks 7
.Sh "BUGS"
.Nm fz
is still very immature. I'm sure it contains bugs and it is certainly missing features.


CHANGELOG.md
# CHANGELOG

## v0.0.22

- Add installation instructions for Alpine Linux.
- Fix logic of the `task/disk_free` task.
- Fix errors in the `task/md_degraded` task.
- Introduce `--pidfile` flag to `fz`.
- Add `task/btrfs_scrub_errors`.
- Add `task/btrfs_scrub_time`.
- Add `task/btrfs_snapshot_age`.
- Improve the output of `notifier/email`.

## v0.0.21

- Support emitting output from tasks into notifications.
- Add `OPEN_GATES` to the environment of notifiers.
- Add `task/file_exists` plugin.
- Add `task/file_max_age` plugin.
- Add `task/loadavg` plugin.
- Add `task/mount` plugin.
- Add `task/swap_free` plugin.
- Add `task/tls_expiration` plugin.
- Modify the `task/disk_free` plugin to use kb, rather than percentage.
- Add tags to messages from `ntfy`.
- Improve timeout handling of task plugins.

## v0.0.20

- Many improvements to the docs and the interface.
- Removed dependency on logrus so that it can cross-compile for OpenBSD 7.5.
- Support gatesets to combine ANDing or ORing the results of gates.
- Improvements to the configtest
- Add `task/fz_errors` plugin
- Add `LAST_OK`, `LAST_FAIL` and `LAST_NOTIFICATION` to the environment of the gates.
- Support delayed notifications.
- Support re-raising notifications.
- Support detecting when a task is flapping.

## v0.0.19

- Add fzctl to check the daemon state
- Add scripts for testing HTTP
- Slight improvements to the man pages

## v0.0.18

- Release a tarball of plugins

## v0.0.17

- Test a release

## v0.0.16

- Simplify the release logic

## v0.0.15

- Improvements to the generated OpenBSD artifact
- Add a configtest mode.

## v0.0.14

- Remove dependency on goreleaser

## v0.0.13

- Fix crashes on OpenBSD

## v0.0.10

- License it under BSD-2-Clause
- Start a man page

## v0.0.7

- Add option parsing

## v0.0.5

- Another test release


cmd/fz/fz.go
package main

import (
	"fmt"
	"log"
	"os"
	"time"

	"github.com/jonhiggs/flamingzombies/lib/daemon"
	"github.com/jonhiggs/flamingzombies/lib/fz"
	"nullprogram.com/x/optparse"
)

const VERSION = "v0.0.21"

var config fz.Config
var configtestMode = false

func init() {
	if os.Getenv("FZ_CONFIG_FILE") == "" {
		os.Setenv("FZ_CONFIG_FILE", "/etc/flamingzombies.toml")
	}

	if os.Getenv("FZ_DIRECTORY") == "" {
		os.Setenv("FZ_DIRECTORY", "/usr/libexec/flamingzombies")
	}

	if os.Getenv("FZ_LISTEN") == "" {
		os.Setenv("FZ_LISTEN", "127.0.0.1:5891")
	}

	options := []optparse.Option{
		{"config", 'c', optparse.KindRequired},
		{"configtest", 'n', optparse.KindNone},
		{"directory", 'C', optparse.KindRequired},
		{"help", 'h', optparse.KindNone},
		{"loglevel", 'l', optparse.KindRequired},
		{"version", 'V', optparse.KindNone},
		{"pidfile", 'p', optparse.KindRequired},
	}

	results, _, err := optparse.Parse(options, os.Args)
	if err != nil {
		log.Fatal(err)
	}

	for _, result := range results {
		switch result.Long {
		case "config":
			os.Setenv("FZ_CONFIG_FILE", result.Optarg)
		case "configtest":
			configtestMode = true
		case "loglevel":
			os.Setenv("FZ_LOG_LEVEL", result.Optarg)
		case "directory":
			os.Setenv("FZ_DIRECTORY", result.Optarg)
		case "pidfile":
			err := os.WriteFile(result.Optarg, []byte(fmt.Sprintf("%d\n", os.Getpid())), 0644)
			if err != nil {
				fmt.Println(err)
				os.Exit(1)
			}
		case "help":
			usage()
			return
		case "version":
			fmt.Printf("fz %s\n", VERSION)
			os.Exit(0)
		}
	}

	config = fz.ReadConfig()

	if os.Getenv("FZ_LOG_LEVEL") != "" {
		fz.StartLogger(os.Getenv("FZ_LOG_LEVEL"))
	} else {
		fz.StartLogger(config.LogLevel)
	}

	if configtestMode {
		fmt.Println("The configuration is valid")
		os.Exit(0)
	}

	fz.ProcessNotifications()

	if config.Listen() {
		go daemon.Listen(&config)
	}
}

func main() {
	ticker := time.NewTicker(time.Second)
	defer ticker.Stop()

	for {
		select {
		case ts := <-ticker.C:
			for i, t := range config.Tasks {
				if t.Ready(ts) {
					go config.Tasks[i].Run()
				}
			}
		}
	}
}

func usage() {
	fmt.Println("Usage:")
	fmt.Println("  fz [OPTIONS]")
	fmt.Println("")
	fmt.Println("Options:")
	fmt.Println("  -c, --config <file>     Configuration file")
	fmt.Println("  -C, --directory <path>  Change to directory")
	fmt.Println("  -h, --help              This help")
	fmt.Println("  -l, --loglevel <level>  Override the log level")
	fmt.Println("  -n, --configtest        Test validity of the configuration")
	fmt.Println("  -p, --pidfile           The pidfile to write")
	fmt.Println("  -V, --version           Version")
	os.Exit(0)
}
